### PR TITLE
Update event mangement logic to factor in resource type when closing resources

### LINKED ISSUE #3149: Update Event Management configs to specify resource type in the filter expression and values to update
**Describe the feature request.**
The current [event management configuration](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/event/EventWorkflow.kt)  has a list of `UpdateWorkflowValueConfig`s  (a list of values to update) and `resourceFilterExpression` (used to filter resources that will be updated).

Both the `updateValues` and `resourceFilterExpression` will be applied to all the resources defined in the `eventWorkFlow` config. This is an issue since one `eventWorkFlow` can be used to define multiple resources for closure e.g a `Careplan` with related `Tasks` and `Conditions`.

The current configs need to be updated so that it's clear which resource type the `updateValues` and `resourceFilterExpression` are referencing.

**Acceptance criteria**

- [ ] All resources defined in the `eventWorkFlow` config that meet the `resourceFilterExpression` ,if present, are updated with the correct values defined in `updateValues` field

**Implementation plan (For Engineers)**

- [ ] Update the `UpdateWorkflowValueConfig` to have a variable that stores the resource type.
- [ ] Update the [ResourceFilterExpression](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ResourceFilterExpression.kt) to have a variable that stores the resource type.
- [ ] Update the `resourceFilterExpression` variable to take a list of `ResourceFilterExpression`s
- [ ] Make the relevant changes to the current Event Management logic that will arise due to the aforementioned changes.
- [ ] Update relevant tests.

Sample of the updated values 
```
 "updateValues": [
                  {
                    "jsonPathExpression": "Task.status",
                    "value": "cancelled",
                    "resourceType": "Task"
                  },
                  {
                    "jsonPathExpression": "CarePlan.status",
                    "value": "completed",
                    "resourceType": "CarePlan"
                  },
                  {
                    "jsonPathExpression": "ServiceRequest.status",
                    "value": "revoked",
                    "resourceType": "ServiceRequest"
                  },
                  {
                    "jsonPathExpression": "Condition.clinicalStatus.coding[0].code",
                    "value": "370996005",
                    "resourceType": "Condition"
                  },
                  {
                    "jsonPathExpression": "Condition.clinicalStatus.coding[0].system",
                    "value": "http://www.snomed.org/",
                    "resourceType": "Condition"
                  },
                  {
                    "jsonPathExpression": "Condition.clinicalStatus.coding[0].display",
                    "value": "resolved",
                    "resourceType": "Condition"
                  }
                ],
                "resourceFilterExpressions": [
                  {
                    "conditionalFhirPathExpressions": [
                      "Task.status != 'completed'"
                    ],
                    "resourceType": "Task"
                  },
                  {
                    "conditionalFhirPathExpressions": [
                      "ServiceRequest.status != 'completed'"
                    ],
                    "resourceType": "ServiceRequest"
                  }
                ]
```

Here is a sample of a complete `eventWorkFlow` config

```
{"eventWorkflows": [
              {
                "eventType": "RESOURCE_CLOSURE",
                "triggerConditions": [
                  {
                    "eventResourceId": "carePlanToBeClosed",
                    "conditionalFhirPathExpressions": [
                      "true"
                    ]
                  },
                  {
                    "eventResourceId": "referralServiceRequestToBeClosed",
                    "conditionalFhirPathExpressions": [
                      "%resource.entry.where(resource is QuestionnaireResponse).resource.where(questionnaire = 'Questionnaire/dc-remove-patient').exists() and %resource.entry.where(resource is QuestionnaireResponse).resource.repeat(item).where(linkId = 'reason-for-removal' and (answer.value.code = 'died' or answer.value.code = 'moved-away')).exists()"
                    ]
                  }
                ],
                "eventResources": [
                  {
                    "id": "carePlanToBeClosed",
                    "resource": "CarePlan",
                    "configRules": [
                      {
                        "name": "patientId",
                        "condition": "true",
                        "actions": [
                          "data.put('patientId', fhirPath.extractValue(Patient, 'Patient.id').contains('Patient') ? fhirPath.extractValue(Patient, 'Patient.id') : 'Patient/' +  fhirPath.extractValue(Patient, 'Patient.id'))"
                        ]
                      }
                    ],
                    "dataQueries": [
                      {
                        "paramName": "instantiates-canonical",
                        "filterCriteria": [
                          {
                            "dataType": "REFERENCE",
                            "value": "PlanDefinition/dc-diabetes-screening-intervention"
                          }
                        ]
                      },
                      {
                        "paramName": "subject",
                        "filterCriteria": [
                          {
                            "dataType": "REFERENCE",
                            "computedRule": "patientId"
                          }
                        ]
                      }
                    ],
                    "relatedResources": [
                      {
                        "resource": "Task",
                        "searchParameter": "based-on"
                      }
                    ]
                  },
                  {
                    "id": "referralServiceRequestToBeClosed",
                    "resource": "ServiceRequest",
                    "configRules": [
                      {
                        "name": "patientId",
                        "condition": "true",
                        "actions": [
                          "data.put('patientId', fhirPath.extractValue(Patient, 'Patient.id'))"
                        ]
                      }
                    ],
                    "dataQueries": [
                      {
                        "paramName": "code",
                        "filterCriteria": [
                          {
                            "dataType": "CODEABLECONCEPT",
                            "value": {
                              "system": "http://snomed.info/sct",
                              "code": "44383000"
                            }
                          }
                        ]
                      },
                      {
                        "paramName": "subject",
                        "filterCriteria": [
                          {
                            "dataType": "REFERENCE",
                            "computedRule": "patientId"
                          }
                        ]
                      }
                    ]
                  }
                ],
                "updateValues": [
                  {
                    "jsonPathExpression": "Task.status",
                    "value": "cancelled",
                    "resourceType": "Task"
                  },
                  {
                    "jsonPathExpression": "CarePlan.status",
                    "value": "completed",
                    "resourceType": "CarePlan"
                  },
                  {
                    "jsonPathExpression": "ServiceRequest.status",
                    "value": "revoked",
                    "resourceType": "ServiceRequest"
                  },
                  {
                      "jsonPathExpression": "Condition.clinicalStatus.coding[0].code",
                      "value": "370996005",
                      "resourceType": "Condition"
                    },
                    {
                      "jsonPathExpression": "Condition.clinicalStatus.coding[0].system",
                      "value": "http://www.snomed.org/",
                      "resourceType": "Condition"
                    },
                    {
                      "jsonPathExpression": "Condition.clinicalStatus.coding[0].display",
                      "value": "resolved",
                      "resourceType": "Condition"
                    }
                ],
                "resourceFilterExpressions": [
                  {
                    "conditionalFhirPathExpressions": [
                      "Task.status != 'completed'"
                    ],
                    "resourceType": "Task"
                  },
                  {
                    "conditionalFhirPathExpressions": [
                      "ServiceRequest.status != 'completed'"
                    ],
                    "resourceType": "ServiceRequest"
                  }
                ]
              }
            ]
}
```

### PR DESCRIPTION
**Pull request by Rkareko**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #3149

**Engineer Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 
- [ ] I have checked that this PR does NOT introduce **breaking changes** that require an update to **_Content_** and/or **_Configs_**? _If it does add a sample here or a link to exactly what changes need to be made to the content._

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

**Comment by Rkareko** [comment]
Sample of the updated values 
```
      "updateValues": [
        {
          "jsonPathExpression": "Task.status",
          "value": "cancelled",
          "resourceType": "Task"
        },
        {
          "jsonPathExpression": "CarePlan.status",
          "value": "completed",
          "resourceType": "CarePlan"
        },
        {
          "jsonPathExpression": "ServiceRequest.status",
          "value": "revoked",
          "resourceType": "ServiceRequest"
        }
      ],
      "resourceFilterExpressions": [
        {
          "conditionalFhirPathExpressions": "Task.status != 'completed'",
          "resourceType": "Task"
        },
        {
          "conditionalFhirPathExpressions": "ServiceRequest.status != 'completed'",
          "resourceType": "ServiceRequest"
        }
      ]
```

Here is a sample of a complete `eventWorkFlow` config

```
{
  "eventWorkflows": [
    {
      "eventType": "RESOURCE_CLOSURE",
      "triggerConditions": [
        {
          "eventResourceId": "carePlanToBeClosed",
          "conditionalFhirPathExpressions": [
            "true"
          ]
        },
        {
          "eventResourceId": "referralServiceRequestToBeClosed",
          "conditionalFhirPathExpressions": [
            "%resource.entry.where(resource is QuestionnaireResponse).resource.where(questionnaire = 'Questionnaire/dc-remove-patient').exists() and %resource.entry.where(resource is QuestionnaireResponse).resource.repeat(item).where(linkId = 'reason-for-removal' and (answer.value.code = 'died' or answer.value.code = 'moved-away')).exists()"
          ]
        }
      ],
      "eventResources": [
        {
          "id": "carePlanToBeClosed",
          "resource": "CarePlan",
          "configRules": [
            {
              "name": "patientId",
              "condition": "true",
              "actions": [
                "data.put('patientId', fhirPath.extractValue(Patient, 'Patient.id'))"
              ]
            }
          ],
          "dataQueries": [
            {
              "paramName": "instantiates-canonical",
              "filterCriteria": [
                {
                  "dataType": "REFERENCE",
                  "value": "PlanDefinition/dc-diabetes-screening-intervention"
                }
              ]
            },
            {
              "paramName": "subject",
              "filterCriteria": [
                {
                  "dataType": "REFERENCE",
                  "computedRule": "patientId"
                }
              ]
            }
          ],
          "relatedResources": [
            {
              "resource": "Task",
              "searchParameter": "based-on"
            }
          ]
        },
        {
          "id": "referralServiceRequestToBeClosed",
          "resource": "ServiceRequest",
          "configRules": [
            {
              "name": "patientId",
              "condition": "true",
              "actions": [
                "data.put('patientId', fhirPath.extractValue(Patient, 'Patient.id'))"
              ]
            }
          ],
          "dataQueries": [
            {
              "paramName": "code",
              "filterCriteria": [
                {
                  "dataType": "CODEABLECONCEPT",
                  "value": {
                    "system": "http://snomed.info/sct",
                    "code": "44383000"
                  }
                }
              ]
            },
            {
              "paramName": "subject",
              "filterCriteria": [
                {
                  "dataType": "REFERENCE",
                  "computedRule": "patientId"
                }
              ]
            }
          ]
        }
      ],
      "updateValues": [
        {
          "jsonPathExpression": "Task.status",
          "value": "cancelled",
          "resourceType": "Task"
        },
        {
          "jsonPathExpression": "CarePlan.status",
          "value": "completed",
          "resourceType": "CarePlan"
        },
        {
          "jsonPathExpression": "ServiceRequest.status",
          "value": "revoked",
          "resourceType": "ServiceRequest"
        }
      ],
      "resourceFilterExpressions": [
        {
          "conditionalFhirPathExpressions": "Task.status != 'completed'",
          "resourceType": "Task"
        },
        {
          "conditionalFhirPathExpressions": "ServiceRequest.status != 'completed'",
          "resourceType": "ServiceRequest"
        }
      ]
    }
  ]
}
```

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/event/EventWorkflow.kt [hunk 1]
```diff
@@ -32,24 +32,27 @@ data class EventWorkflow(
...
   val updateValues: List<UpdateWorkflowValueConfig> = emptyList(),
-  val resourceFilterExpression: ResourceFilterExpression? = null,
+  val resourceFilterExpressions: List<ResourceFilterExpression>? = null,
 ) : java.io.Serializable, Parcelable
 
 @Serializable
 data class UpdateWorkflowValueConfig(
   val jsonPathExpression: String,
   val value: JsonElement,
+  val resourceType: String,
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/event/EventWorkflow.kt:42]:
```suggestion
  val resourceType: ResourceType,
```

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt [hunk 2]
```diff
@@ -824,29 +845,31 @@ constructor(
...
+              set(updateExpression.jsonPathExpression, updateValue)
+            }
+            if (
+              updateExpression.jsonPathExpression.startsWith(
+                resource.resourceType.name,
+                ignoreCase = true,
+              ) && updateExpression.value != null
+            ) {
+              set(
+                updateExpression.jsonPathExpression.replace(resource.resourceType.name, "\$"),
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/DefaultRepository.kt:868]:
```suggestion
                updateExpression.jsonPathExpression.replace(resource.resourceType.name, "\$", ignoreCase = true),
```

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ResourceFilterExpression.kt [hunk 3]
```diff
@@ -16,10 +16,14 @@
...
+import kotlinx.parcelize.Parcelize
 import kotlinx.serialization.Serializable
 
 @Serializable
+@Parcelize
 data class ResourceFilterExpression(
   val conditionalFhirPathExpressions: List<String>,
   val matchAll: Boolean = true,
-) : java.io.Serializable
+  val resourceType: String = "",
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ResourceFilterExpression.kt:28]:
```suggestion
  val resourceType: ResourceType,
```

**Review by ellykits — CHANGES_REQUESTED** [review]

**Comment by Rkareko** [comment]
App crashes when loading Profile screen. Do not merge till this is fixed.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/event/EventWorkflow.kt [hunk 4]
```diff
@@ -32,24 +33,27 @@ data class EventWorkflow(
...
   val updateValues: List<UpdateWorkflowValueConfig> = emptyList(),
-  val resourceFilterExpression: ResourceFilterExpression? = null,
+  val resourceFilterExpressions: List<ResourceFilterExpression>? = null,
 ) : java.io.Serializable, Parcelable
 
 @Serializable
 data class UpdateWorkflowValueConfig(
   val jsonPathExpression: String,
   val value: JsonElement,
+  val resourceType: ResourceType = ResourceType.Task,
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/event/EventWorkflow.kt:43]: Is defaulting to `Task` resource necessary?

**Review by ellykits — COMMENTED** [review]
This LGTM? left a single comment.

**Comment by ellykits** [comment]
> App crashes when loading Profile screen. Do not merge till this is fixed.

@Rkareko This could be related to the updated configs. The existing configs need to be updated for the rest of the apps using this feature.

**Comment by ellykits** [comment]
Update the docs, then we can have this merged.

**Review by ellykits — APPROVED** [review]

**Pull request merged**